File: mobile/packages/core/test/sealed_box_test.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:offlinepay_core/offlinepay_core.dart';
import 'package:test/test.dart';

void main() {
  test('round trip in Dart', () async {
    final recip = await generateSealedBoxKeyPair();
    final msg = utf8.encode('secret transaction blob');
    final ct = await sealAnonymous(recip.publicKey, msg);
    expect(ct.length, equals(msg.length + sealedBoxOverhead));
    final pt = await openAnonymous(recip.publicKey, recip.keyPair, ct);
    expect(pt, equals(msg));
  });

  test('fresh ephemeral keypair per seal', () async {
    final recip = await generateSealedBoxKeyPair();
    final msg = utf8.encode('same');
    final a = await sealAnonymous(recip.publicKey, msg);
    final b = await sealAnonymous(recip.publicKey, msg);
    expect(a, isNot(equals(b)));
  });

  test('tamper fails', () async {
    final recip = await generateSealedBoxKeyPair();
    final ct = await sealAnonymous(recip.publicKey, utf8.encode('hi'));
    ct[ct.length - 1] ^= 0xff;
    expect(
      () => openAnonymous(recip.publicKey, recip.keyPair, ct),
      throwsA(isA<Object>()),
    );
  });

  test('wrong recipient fails', () async {
    final a = await generateSealedBoxKeyPair();
    final b = await generateSealedBoxKeyPair();
    final ct = await sealAnonymous(a.publicKey, utf8.encode('hi'));
    expect(
      () => openAnonymous(b.publicKey, b.keyPair, ct),
      throwsA(isA<Object>()),
    );
  });

  test('short ciphertext rejected', () async {
    final recip = await generateSealedBoxKeyPair();
    expect(
      () => openAnonymous(recip.publicKey, recip.keyPair, Uint8List.fromList(utf8.encode('too short'))),
      throwsA(isA<Object>()),
    );
  });
}

File: backend/internal/crypto/sealed_box_test.go
package crypto

import (
	"bytes"
	"testing"
)

func TestSealedBoxRoundTrip(t *testing.T) {
	pub, priv, err := GenerateSealedBoxKeyPair()
	if err != nil {
		t.Fatal(err)
	}
	msg := []byte("secret transaction blob")
	ct, err := SealAnonymous(pub, msg)
	if err != nil {
		t.Fatal(err)
	}
	if len(ct) != len(msg)+SealedBoxOverhead {
		t.Fatalf("overhead mismatch: got %d want %d", len(ct)-len(msg), SealedBoxOverhead)
	}
	pt, err := OpenAnonymous(pub, priv, ct)
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(pt, msg) {
		t.Fatalf("plaintext mismatch: got %q want %q", pt, msg)
	}
}

func TestSealedBoxWrongRecipient(t *testing.T) {
	pub1, _, _ := GenerateSealedBoxKeyPair()
	pub2, priv2, _ := GenerateSealedBoxKeyPair()
	ct, _ := SealAnonymous(pub1, []byte("hi"))
	if _, err := OpenAnonymous(pub2, priv2, ct); err == nil {
		t.Fatal("expected decrypt failure with wrong recipient")
	}
}

func TestSealedBoxTamper(t *testing.T) {
	pub, priv, _ := GenerateSealedBoxKeyPair()
	ct, _ := SealAnonymous(pub, []byte("hi"))
	ct[len(ct)-1] ^= 0xff
	if _, err := OpenAnonymous(pub, priv, ct); err == nil {
		t.Fatal("expected tamper detection")
	}
}

func TestSealedBoxShortCiphertext(t *testing.T) {
	pub, priv, _ := GenerateSealedBoxKeyPair()
	if _, err := OpenAnonymous(pub, priv, []byte("too short")); err == nil {
		t.Fatal("expected error on short ciphertext")
	}
}

func TestSealedBoxFreshEphemeral(t *testing.T) {
	// Sealing the same plaintext twice must produce different ciphertexts
	// because of the fresh ephemeral keypair per call.
	pub, _, _ := GenerateSealedBoxKeyPair()
	a, _ := SealAnonymous(pub, []byte("same"))
	b, _ := SealAnonymous(pub, []byte("same"))
	if bytes.Equal(a, b) {
		t.Fatal("expected different ciphertexts with fresh ephemeral keypairs")
	}
}
